File: greedy/activity-selection/activity-selection.ts
// Greedy: Activity Selection (maximum non-overlapping activities)

type Activity = [start: number, end: number];

function activitySelection(activities: Activity[]): number {
  const sorted = [...activities].sort((a, b) => a[1] - b[1]);

  let count = 1;
  let lastEnd = sorted[0][1];

  for (let i = 1; i < sorted.length; i++) {
    const [start, end] = sorted[i];
    if (start >= lastEnd) {
      count++;
      lastEnd = end;
    }
  }

  return count;
}

function main(): void {
  const activities: Activity[] = [
    [1, 3],
    [2, 5],
    [4, 6],
    [6, 7],
    [5, 9],
    [8, 9],
  ];

  console.log(`Max activities: ${activitySelection(activities)}`);
}

main();

File: greedy/activity-selection/activity-selection.py
"""Greedy: Activity Selection (maximum non-overlapping activities)"""


def activity_selection(activities: list[tuple[int, int]]) -> int:
    activities = sorted(activities, key=lambda a: a[1])

    count = 1
    last_end = activities[0][1]

    for start, end in activities[1:]:
        if start >= last_end:
            count += 1
            last_end = end

    return count


if __name__ == "__main__":
    activities = [(1, 3), (2, 5), (4, 6), (6, 7), (5, 9), (8, 9)]
    print(f"Max activities: {activity_selection(activities)}")
